### ISSUE TITLE
Expose the `EventRead` metadata along with a value

### ISSUE THREAD

**Issue by EronWright** [pravega/flink-connectors#180]
**Problem description**
Make it possible to create a `FlinkPravegaReader` that provides event metadata (e.g. `EventPosition`) for each event.    This would facilitate the development of jobs that care about the stream position of the event data, e.g. for indexing purposes.

**Suggestions for an improvement**
Normally, the data stream that is produced is of type `DataStream<T>` to convey the deserialized event data of type `T`.   To also convey metadata implies a data stream of type `DataStream<EventRead<T>>`.   For example:

```java
FlinkPravegaReader<EventRead<Double>> reader = ...;
DataStream<EventRead<Double>> datastream = env.addSource(reader);
datastream.map(e -> {
  assert e.isEvent();
  LOG.info("event: data={}, position={}", e.getEvent(), e.getEventPosition());
});
```

The `EventRead` class might not be ideal; we could introduce a similar type based on discussion.

One complication is that the `DeserializationSchema<T>` that is provided to the reader is typed to the event data type.   Some cleverness in the builder would be needed to disambiguate the deserialized type  `<T>` from the data stream type `EventRead<T>`.    May also need to leverage Flink's `OutputTypeConfigurable`.

**Comment by crazyzhou** [comment]
I started a design document here, please have a review @EronWright @claudiofahey : https://docs.google.com/document/d/1Dp8KYA6N81869dFjhD7w3z2hXd2QkU-vS1DNkz5aRC4/edit?usp=sharing

**Comment by crazyzhou** [comment]
The fix #305 is to provide a method `deserializeWithMetadata` for user to override that can leverage `EventRead` in the deserialization.
